### PR TITLE
SQL: fuse projections and tables

### PR DESCRIPTION
**Pull request by webmiche**
This PR adds the fusion of tables with subsequent (simple) projections. In order t do that, it introduces the `partial_table_scan` operation, an operation similar to the `full_table_scan`, but with col_names to define which columns should be loaded.

### CODE REVIEW

File: experimental/sql/tools/rel_opt.py
Comment (webmiche) [inline experimental/sql/tools/rel_opt.py:38]: Looking for a better name for this :)

File: experimental/sql/src/impl_to_iterators.py
Comment (ingomueller-net) [inline experimental/sql/src/impl_to_iterators.py:270]: I suppose that this assumes that the table mapping has been populated with column batch arguments that have the corresponding projection applied, right? Just to be sure: the logic that applies these projections isn't part of this commit, yet, right? Can you explain this mechanism in comments?

Comment (ingomueller-net) [reply experimental/sql/src/impl_to_iterators.py:270]: Oh, I see, `get_batch_and_name_list` applies the projections. Can you explain that here, where get_batch_and_name_list` does the projection, and in the documentation of that function?

File: experimental/sql/tools/rel_opt.py
Comment (ingomueller-net) [reply experimental/sql/tools/rel_opt.py:38]: `-fuse-proj-into-scan`?

**Review by ingomueller-net — APPROVED** [review]
Does this work end-to-end already? (If yes, shouldn't there be an end-to-end test?)

Otherwise, LGTM!

**Comment by webmiche** [comment]
This works end-to-end and it is what I am currently using on my evaluation branch. I was not sure about an end-to-end test as we did not introduce any tests covering multiple layers so far, so I did not want to do this with this PR. However, I feel as tough it might make sense to introduce a full vertical test (with execution) at some point (maybe directly after this PR). For that, however, we will need things that are in #553 plus some more stuff on my evaluation branch, so I want to wait for that to be merged. What do you think?

File: experimental/sql/dialects/rel_impl.py
Comment (webmiche) [reply experimental/sql/dialects/rel_impl.py]: I agree and I implemented the approach with an optional attribute.

File: experimental/sql/src/impl_to_iterators.py
Comment (webmiche) [reply experimental/sql/src/impl_to_iterators.py]: Actually, it's not really get_batch_and_name_list that does the projection since we want to project on to SSAValues (in this case BlockArguments matching the right table). And these block arguments can only be built later on.

So this function really only returns batchtypes and names and does not do any mapping.

I will add some comments to make this clearer.

**Comment by webmiche** [comment]
Not sure why these CI jobs are failing. Maybe @ingomueller-net or @nicolasvasilache have an idea?

File: experimental/sql/src/impl_to_iterators.py
Comment (ingomueller-net) [inline experimental/sql/src/impl_to_iterators.py:271]: Maybe you can factor out `table_name` (or similar), which is the only difference between the two cases.

File: experimental/sql/dialects/rel_impl.py
Comment (ingomueller-net) [reply experimental/sql/dialects/rel_impl.py]: OK, cools, looks good!

File: experimental/sql/src/impl_to_iterators.py
Comment (ingomueller-net) [reply experimental/sql/src/impl_to_iterators.py]: Aha, right, the actual projection is done even earlier, completely outside of the plan. `get_batch_and_name_list` populates the column batches that are accessed by some scan op, some of which may be projected before the call into the plan. The new comments are better, but it doesn't mention yet that there is this convention between the call site and the plan. Can you elaborate on that?

**Review by ingomueller-net — COMMENTED** [review]
Cool, thanks for the update! Only very minor comments, now. LGTM!

Aha, of course, the target dialect isn't ready in main yet 😛 Sure, let's add end-to-end tests once that is done.

**Comment by ingomueller-net** [comment]
> Not sure why these CI jobs are failing. Maybe @ingomueller-net or @nicolasvasilache have an idea?

The tests failed with some connection error to Github. My best guess is that this is transient. I restarted the job -- let's see if this solves the problem...

File: experimental/sql/src/impl_to_iterators.py
Comment (webmiche) [reply experimental/sql/src/impl_to_iterators.py]: Are you referring to the projection of the columnar batch itself? I am not sure whether this is actually the right approach. We might have to think about changing your scan_columnar_batch at some point to also take col names and only builds tuples from these columns.

I changed the comment again :)

Comment (ingomueller-net) [reply experimental/sql/src/impl_to_iterators.py]: You haven't pushed that comment yet, have you?

Comment (webmiche) [reply experimental/sql/src/impl_to_iterators.py]: Yes, indeed. It's there now :)

File: experimental/sql/src/impl_to_iterators.py
Comment (ingomueller-net) [inline experimental/sql/src/impl_to_iterators.py]: Only realizing this now: Is there any reason not to make "partial" scans an option of the full scan in this dialect as well?

File: experimental/sql/src/impl_to_iterators.py
Comment (ingomueller-net) [inline experimental/sql/src/impl_to_iterators.py]: I think the documentation of this function (`Scans over all the operations ...`) should be updated to explain what happens with partial scans/projections.

File: experimental/sql/src/impl_to_iterators.py
Comment (ingomueller-net) [reply experimental/sql/src/impl_to_iterators.py]: Thanks for the update! I still don't understand where exactly the projection happens, though. As I understand it, the lowered code can use some SSA values with tables where the projection has already been applied (i.e., a columnar batch with only a subset of the original columns) and the table mapping contains the names of those SSA values. But who creates these values?

File: experimental/sql/src/impl_to_iterators.py
Comment (webmiche) [reply experimental/sql/src/impl_to_iterators.py]: This comment is out of date, right?

File: experimental/sql/src/impl_to_iterators.py
Comment (webmiche) [reply experimental/sql/src/impl_to_iterators.py]: This comment is out of date, right?

**Comment by webmiche** [comment]
I feel that this can be merged, if the two comments I just addressed are actually out of date (@ingomueller-net ?).

**Comment by ingomueller-net** [comment]
Yes, sorry, I must have looked at some earlier commit when I made these comments. Good to go!

**Pull request merged**
